kong/types.go

```diff
@@ -267,3 +270,84 @@ type Role struct {
 	Comment   *string `json:"comment,omitempty" yaml:"comment,omitempty"`
 	IsDefault *bool   `json:"is_default,omitempty" yaml:"is_default,omitempty"`
 }
+
+// PermissionRoleID represents a Role identifier within a permission struct
+type PermissionRoleID struct {
+	ID *string `json:"id,omitempty" yaml:"id,omitempty"`
+}
+
+// EndpointPermission represents an RBAC Endpoint Permisson in Kong Enterprise
+type EndpointPermission struct {
+	CreatedAt *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	Workspace *string           `json:"workspace,omitempty" yaml:"workspace,omitempty"`
+	Endpoint  *string           `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+	Actions   []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative  *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role      *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+	Comment   *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+}
+
+// MarshalJSON marshals an endpoint permission into a suitable form for the Kong admin API
+func (e *EndpointPermission) MarshalJSON() ([]byte, error) {
+	type ep struct {
+		CreatedAt *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+		Workspace *string           `json:"workspace,omitempty" yaml:"workspace,omitempty"`
+		Endpoint  *string           `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+		Actions   *string           `json:"actions,omitempty" yaml:"actions,omitempty"`
+		Negative  *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+		Role      *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+		Comment   *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+	}
+	var actions []string
+	for _, action := range e.Actions {
+		actions = append(actions, *action)
+	}
+	return json.Marshal(&ep{
+		CreatedAt: e.CreatedAt,
+		Workspace: e.Workspace,
+		Endpoint:  e.Endpoint,
+		Actions:   String(strings.Join(actions, ",")),
+		Comment:   e.Comment,
+	})
+}
+
+// EntityPermission represents an RBAC Entity Permission in Kong Enterprise
+type EntityPermission struct {
+	CreatedAt  *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	EntityID   *string           `json:"entity_id,omitempty" yaml:"entity_id,omitempty"`
+	EntityType *string           `json:"entity_type,omitempty" yaml:"entity_type,omitempty"`
+	Actions    []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative   *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role       *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
```

Comment:
Same question here, why can't this use `RBACRole`?

Reply:
I suppose we can -- marshaling would just omit the empty fields.